test/e2e/storage/non_csi_volumes.go

```diff
@@ -0,0 +1,758 @@
+	"hostpath":                   initHostpathDriver,
+	"hostpathSymlink":            initHostpathSymlinkDriver,
+	"emptydir":                   initEmptydirDriver,
+}
+
+var _ = utils.SIGDescribe("Non-CSI", func() {
+	f := framework.NewDefaultFramework("volumes")
+
+	var (
+		cs     clientset.Interface
+		ns     *v1.Namespace
+		node   v1.Node
+		config framework.VolumeTestConfig
+	)
+
+	BeforeEach(func() {
+		cs = f.ClientSet
+		ns = f.Namespace
+		nodes := framework.GetReadySchedulableNodesOrDie(f.ClientSet)
+		node = nodes.Items[rand.Intn(len(nodes.Items))]
+		config = framework.VolumeTestConfig{
+			Namespace:      ns.Name,
+			Prefix:         "volume",
+			ClientNodeName: node.Name,
+			ServerNodeName: node.Name,
+		}
+	})
+
+	for driverName, initDriver := range testDrivers {
+		curDriverName := driverName
+		curInitDriver := initDriver
+
+		Context(fmt.Sprintf("using driver: %s", curDriverName), func() {
+			var (
+				driver testDriver
+			)
+
+			BeforeEach(func() {
+				driver = curInitDriver(f, config, node)
+				driver.createDriver()
+			})
+
+			AfterEach(func() {
+				driver.cleanupDriver()
+			})
+
+			It("should be mountable", func() {
+				tests := createVolumeTest(driver)
+				for _, t := range tests {
+					defer framework.VolumeTestCleanup(f, t.config)
+					needInjectHTML, ok := needInjectHTML[t.name]
+					if ok && needInjectHTML {
+						volumeTest := t.tests
+						framework.InjectHtml(cs, t.config, volumeTest[0].Volume, volumeTest[0].ExpectedContent)
+					}
+					framework.TestVolumeClient(cs, t.config, t.fsGroup, t.tests)
+				}
+			})
+
+			It("should write files of various sizes, verify size, validate content [Slow]", func() {
+				tests := createVolumeIOTest(driver)
+				for _, t := range tests {
+					err := testVolumeIO(f, cs, t.config, t.volSource, &t.podSec, t.testFile, t.fileSizes)
+					Expect(err).NotTo(HaveOccurred())
+				}
+			})
+
+			It("should show a volume with proper volume mode to a pod [Feature:BlockVolume]", func() {
+				tests := createPersistentVolumeModeTest(driver)
+				for _, t := range tests {
+					testVolumeMode(f, cs, t.config, ns.Name, t.scName, t.volBindMode, t.volMode, t.pvSource, t.isBlockSupported)
+				}
+			})
+
+		})
+
+		for _, test := range subPathTests {
+			currentTest := test
+
+			Context(fmt.Sprintf("using driver: %s(PV: %v), SubPath test",
+				curDriverName, currentTest.isPVtest), func() {
+				var (
+					driver testDriver
+
+					testResource *subPathTestResource
+					testInputs   subPathTestInputs
+				)
+
+				BeforeEach(func() {
+					driver = curInitDriver(f, config, node)
+					driver.createDriver()
+
+					testResource = createSubPathTestResource(f, currentTest, driver)
+					testInputs = createSubPathInputs(f, testResource, node)
+				})
+
+				AfterEach(func() {
+					By("Cleaning up test resources")
+					cleanupSubPathTestResource(f, testInputs, testResource)
+
+					driver.cleanupDriver()
+				})
+
+				testSubPath(&testInputs)
```

Comment:
It seems to me that subpath tests are run twice, subpath.go is the second place.

Reply:
I put both PV tests and volume tests in the same loop, so testSubPath will be called twice for them.

In other word, I generalize what was done in nfs and nfsPVC and apply it to all other drivers.
So, drivers which return both PV and volume, like nfs, cephFS, rbd, and iSCSI, are tested for both.
Also, tests for drivers which don't return either of them will be skipped.